graph/pull.go

```diff
@@ -635,24 +638,38 @@ func (s *TagStore) pullV2Tag(r *registry.Session, out io.Writer, endpoint *regis
 			return nil
 		}
 
+		// Try to lock the image to make sure we're the only one downloading.
+		// Note: If we successfully lock (pullAlreadInProgress = false), then we are
+		// also responsible for removing ourselves from the pool, in success or failure.
+		c, err := s.poolAdd("pull", "img:"+img.ID)
+		downloads[i].pullAlreadyInProgress = err != nil
+		downloads[i].waitChannel = c
+
 		downloads[i].err = make(chan error)
 		go func(di *downloadInfo) {
 			di.err <- downloadFunc(di)
 		}(&downloads[i])
 	}
 
+	var lastErr error
 	var tagUpdated bool
 	for i := len(downloads) - 1; i >= 0; i-- {
 		d := &downloads[i]
 		if d.err != nil {
 			if err := <-d.err; err != nil {
-				return false, err
+				logrus.Debugf("Failed to download %s with err %s", d.img.ID, err)
+				lastErr = err
 			}
 		}
-		if d.downloaded {
-			// if tmpFile is empty assume download and extracted elsewhere
+
+		if d.tmpFile != nil {
```

Comment:
This seemed to be a bug before; if all downloads were not successful, we would leak the temp downloaded files.